R/check_structure.R

```diff
@@ -0,0 +1,135 @@
+    gsub(
+      pattern = "(s\\wp)-\\d*-(\\w{2})", replacement = "template_\\1_\\2",
+      protocol_code
+    )
+  path_to_template <-
+    system.file(
+      file.path("rmarkdown", "templates", template_name, "skeleton"),
+      package = "protocolhelper")
+  files_protocol <- list.files(path = path_to_protocol)
+  files_template <- list.files(path = path_to_template)
+
+  # check if file(name)s from template are conserved
+  files_template_i <-
+    c(files_template[files_template != "skeleton.Rmd"], "index.Rmd")
+  difffiles <- files_template_i[!files_template_i %in% files_protocol]
+  difffiles <- difffiles[!grep("^\\d{2}_appendices.Rmd", difffiles)]
+  difffiles <- difffiles[!grep("^\\d{2}_subprocotols.Rmd", difffiles)]
+  assert_that(
+    length(difffiles) == 0,
+    msg = paste(protocol_code, "lacks file(s)", difffiles)
+  )
+
+  for (file in files_protocol[grepl(".Rmd$", files_protocol)]) {
+
+    # check if chunks in Rmd files are correct
+    rmd <- readLines(file.path(path_to_protocol, file))
+    start_chunk <- grep("^```\\{r.*}", rmd)
+    end_chunk <- grep("^```[:space:]?$", rmd)
+    assert_that(
+      length(start_chunk) == length(end_chunk),
+      all(start_chunk < end_chunk),
+      msg = paste(protocol_code, ", file", file, "has a problem with R chunks")
+    )
+    for (i in rev(seq_along(start_chunk))) {
+      rmd <- c(
+        head(rmd, start_chunk[i] - 1),
+        "{r code chunk}",
+        tail(rmd, length(rmd) - end_chunk[i])
+      )
+    }
+
+    # check headings general
+    headings <- rmd[grepl("^[[:space:]]?#", rmd)]
+    assert_that(
+      !any(grepl("^[[:space:]]+#", headings)),
+      msg =
+        paste(protocol_code, ", file", file,
+              "Headings have to start with a '#'")
+    )
+    assert_that(
+      !any(grepl("[[:space:]]+$", headings)),
+      msg = paste(
+        protocol_code,
+        "Whitespace at the end of a heading is not allowed",
+        paste(headings[grepl("[[:space:]]+$", headings)], collapse = "\n"),
+        sep = "\n"
+      )
+    )
+
+    # compare headings with template
+    if (file %in% files_template) {
+      template <- readLines(file.path(path_to_template, file))
+      headings_template <- template[grepl("^[[:space:]]?#", template)]
+      headings_template <-
+        headings_template[!grepl("^### Subtit", headings_template)]
+      assert_that(
+        all(headings_template %in% headings),
+        msg =
+          paste(protocol_code, "Heading(s)",
+                headings_template[!headings_template %in% headings],
+                "lack(s) in file", file)
+      )
+      assert_that(
+        all(headings[grepl("^# .*", headings)] %in%
+              headings_template[grepl("^# .*", headings_template)]),
+        msg = paste(protocol_code, "Heading 1",
+               !headings[grepl("^# .*", headings)] %in%
+                 headings_template[grepl("^# .*", headings_template)],
+               "is not allowed in file", file)
+      )
```

Comment:
Should the message return the title of the heading that is not allowed? I think currently the R code in the message returns a logical vector instead of the heading title